Refactor fruit catching logic

ekrctb opened the pull request:
To make the future implementation of caught object pooling easier.

- Move fruit catching logic to `Catcher`.
  - I think it is more logical.
- Simplify fruit catching code.
  - No longer wait for DHO loading until the fruit is dropped.
- Found a possibly-unintended behavior in fruit stacking logic
  - `f.DrawSize.X / 2 * f.Scale.X` was always equal to constant `CatchHitObject.OBJECT_RADIUS / 2`, because fruit scaling is done in the child `ScalingContainer`.
- Simplify fruit dropping code.
  - There was a repeated `Remove` call which has a quadratic time complexity, it should be near-linear time now.
- `CatcherArea` tests are moved to `Catcher` test scene, as the relevant code is moved.
- Add more `Catcher` tests.
  - Found another possibly-unintended behavior: circle size scaling is applied twice to the fruits on the plate.
    - This can be changed when a caught object is no longer a DHO.

ekrctb committed:
Create children in the constructor

ekrctb committed:
Don't delay caught fruit loading

It is not needed anymore because
some code in DCHO is moved from `load` to constructor.

ekrctb committed:
Remove unused method

ekrctb committed:
Use more specific type for caught object

ekrctb committed:
Simplify the fruit stacking code

It is now more clear that
the expression of distance checking is probably unintended (a bug)

ekrctb committed:
Refactor fruit dropping code

- The repeated `Remove` call was quadratic complexity.
  Now it is linear time.

ekrctb committed:
Move more logic to Catcher from CatcherArea

ekrctb committed:
Add more TestSceneCatcher tests

Some tests are moved from TestSceneCatcherArea

peppy commented on osu.Game.Rulesets.Catch/UI/Catcher.cs:
```diff
@@ -479,33 +454,109 @@ private void updateState(CatcherAnimationState state)
+            d.Anchor = Anchor.TopLeft;
+            d.Position = caughtFruitContainer.ToSpaceOfOtherDrawable(d.DrawPosition, droppedObjectTarget);
+
+            animate(d, animation, originalX);
```
I believe you're going to need to re-add these animations after a skin change (since an `updateState` is forced by `DrawableHitObject` on skin change):

![2020-12-03 17 57 15](https://user-images.githubusercontent.com/191335/100986818-0aa44580-3591-11eb-9b0c-3de1629e3c0a.gif)


You can test this easily using the "random skin" hotkey (Ctrl+Shift+R)

peppy reviewed (changes requested):
Code looks a *lot* better than it used to.

ekrctb committed:
Use ApplyCustomUpdateState for dropping transformation

We cannot just apply the transforms because DHO clears transforms when state is updated

ekrctb committed:
Make DroppedObjectAnimation private

bdach reviewed (changes requested):
Generally like this, but some things can still be improved upon.

ekrctb committed:
Fix & clarify catcher tests

ekrctb commented on osu.Game.Rulesets.Catch.Tests/TestSceneCatcher.cs:
```diff
@@ -1,26 +1,192 @@
+            {
+                HyperDashTarget = new Fruit { X = 100 }
+            }));
+            AddStep("miss banana", () => attemptCatch(new Banana()));
```
Nice catch. It was meant to be `TestBananaMissPreservesCatcherState`.

ekrctb committed:
Fix depth of dropped objects

ekrctb committed:
Invert `if`

peppy commented on osu.Game.Rulesets.Catch/UI/Catcher.cs:
```diff
@@ -479,33 +453,109 @@ private void updateState(CatcherAnimationState state)
 
-            void onFruitOnApplyCustomUpdateState(DrawableHitObject o, ArmedState state)
+            // we cannot just apply the transforms because DHO clears transforms when state is updated
+            d.ApplyCustomUpdateState += (o, state) => animate(o, animation, originalX, startTime);
+            if (d.IsLoaded)
+                animate(d, animation, originalX, startTime);
```
another unfortunate case where it would have been great if we could use the `UpdateState` bindable, but i'm guessing we can't due to state changes not always being fired when we expect them to be.

peppy committed:
Merge branch 'master' into caught-object-refactor

peppy committed:
Tidy up code formatting and remove unnecessarily publicly exposed methods

peppy committed:
Move public methods upwards

peppy commented:
@ekrctb I made some code quality changes, can you make sure you're okay with them? Specific attention to the public methods I made private (in case you had intention to use them in the future for testing or otherwise).

ekrctb commented:
@peppy The changes look good to me.

peppy commented:
@bdach want to give this a quick re-review before merge?

bdach commented:
Seems ok from a rapid read-through.
